mdp/utils/covariance.py

```diff
@@ -133,6 +135,119 @@ def fix(self, center=True):
         return cov_mtx, avg, tlen
 
 
+class UnevenlySampledCovarianceMatrix(CovarianceMatrix):
+    """This class stores an empirical covariance matrix that can be updated
+    incrementally. A call to the 'fix' method returns the current state of
+    the covariance matrix, the average and the number of observations, and
+    resets the internal data.
+
+    As compared to the *CovarianceMatrix* class, this class accepts unevenly
+    sampled input in conjunction with a time increment between samples. The
+    covariance matrix is then computed as a (centered) scalar product
+    between functions, that is sampled unevenly, using the trapezoid rule.
+
+    The mean is computed using the trapezoid rule as well.
+
+    Note that the internal sum is a standard __add__ operation.
+    """
+
+    def __init__(self, dtype=None):
+        """Initialize an *UnevenlySampledCovarianceMatrix*.
+
+        :param dtype: Datatype of the input.
+            Default is None. If dtype is not defined, it will be inherited
+            from the first data bunch received by 'update'. 
+        :type dtype: numpy.dtype or str
+        """
+        super(UnevenlySampledCovarianceMatrix,
+              self).__init__(dtype=dtype, bias=True)
+        # we need to keep steps and actual time
+        # values seperately, due to the weighing
+        # which is not required in the evenly sampled case
+        self._steps = 0
+
+    def update(self, x, dt):
+        """Update internal structures.
+
+        Note that no consistency checks are performed on the data (this is
+        typically done in the enclosing node).
+
+        :param x: Timed sequence of random vectors, with samples along the rows
+        and random variables along the colums
+        :type x: numpy.ndarray
+
+        :param dt: Sequence of time increments between random vectors. Must be
+        of length *x.shape[0]-1*.
+        :type dt: numpy.ndarray        
+
+        """
+        if self._cov_mtx is None:
+            self._init_internals(x)
+        # cast input
+        x = mdp.utils.refcast(x, self._dtype)
+        # update the covariance matrix, the average and the number of
+        # observations
+        # the implementation is analogous to the evenly sampled case
+        # values might get big as it is only normalized in fix
+        xcpy = numx.multiply(x[1:, :], numx.sqrt(dt/2.)[:, None])
+        self._cov_mtx += mdp.utils.mult(xcpy.T, xcpy)
+        numx.multiply(x[:-1, :], numx.sqrt(dt/2.)[:, None], out=xcpy)
+        self._cov_mtx += mdp.utils.mult(xcpy.T, xcpy)
+
+        numx.multiply(x[:-1, :], (dt/2.)[:, None], out=xcpy)
+        self._avg += xcpy.sum(axis=0)
+
+        numx.multiply(x[1:, :], (dt/2.)[:, None], out=xcpy)
+        self._avg += xcpy.sum(axis=0)
```

Comment:
The lines
```
numx.multiply(x[:-1, :], (dt/2.)[:, None], out=xcpy)
self._avg += xcpy.sum(axis=0)

numx.multiply(x[1:, :], (dt/2.)[:, None], out=xcpy)
self._avg += xcpy.sum(axis=0)
```
can be optimized as follows (due to linearity):
```
numx.multiply(x[:-1, :], dt[:, None], out=xcpy)
self._avg += xcpy.sum(axis=0)/2.

numx.multiply(x[1:, :], dt[:, None], out=xcpy)
self._avg += xcpy.sum(axis=0)/2.
```
This avoids that `(dt/2.)` allocates a numpy-internal intermediate copy and also saves lots of `/2.` divisions.

Why is `[:, None]` necessary? Wouldn't it be easier if `dt` had `len(dt.shape) == 1`?

Reply:
It is necessary, I believe, to have `dt` applied to all variables along the second dimension of `x`. The downside of dividing by 2 after summing is running the danger of significantly larger value occurring. I adapted it nevertheless, as it still seems to be the more sensible solution. We may want to keep that in mind though (I will include it in the docstrings or comments).

Reply:
I once was also concerned with the much larger value occurring when I looked into the `CovarianceMatrix` class and saw that it adds up all values and only divides by length in the `fix` method. However it doesn't appear to be an issue there, although I am not floating point expert enough to make a thoroughly educated statement. For now I would simply trust the authors of `CovarianceMatrix` that the large value is good enough. All the many intermediate divisions would as well increase the floating point errors. To really tell the difference we would have to test a case where an analytical solution is known.

Reply:
Check out the first entry [here](http://mdp-toolkit.sourceforge.net/additional_utilities.html#additional-utilities).